### PR TITLE
Review ci workflow

### PR DESCRIPTION
**Pull request by crazy-max**

### CODE REVIEW

**Comment by tonistiigi** [comment]
Does this guarantee validations pass before publishing?

**Comment by crazy-max** [comment]
I was thinking doing the same as buildkit/buildx. Do you prefer a single workflow?

**Comment by tonistiigi** [comment]
Yes, I think we should fix these as well. The logic we discussed was to separate push and build steps (with proper caching). Validate can run in parallel with build on PRs, but on releases push only runs if both build and validate were successful.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,34 +1,140 @@
...
+    steps:
       -
         name: Checkout
         uses: actions/checkout@v2
       -
-        name: Validate all
+        name: Cache ${{ matrix.target }}
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.binfmt-cache/${{ matrix.target }}
```
Comment (tonistiigi) [inline .github/workflows/ci.yml:79]: why is the matrix component needed in the path?

Comment (crazy-max) [reply .github/workflows/ci.yml]: Yes we can use a single cache

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,34 +1,140 @@
...
     branches:
       - master
+    tags:
+      - v*
   pull_request:
     branches:
       - master
 
+env:
+  CACHEKEY_MAINLINE_ALL: "mainline-all"
```
Comment (tonistiigi) [inline .github/workflows/ci.yml:15]: looks like these are unused

Comment (crazy-max) [reply .github/workflows/ci.yml]: Yes missed that

**Review by tonistiigi — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -1,34 +1,136 @@
...
+      -
+        # FIXME: Temp fix for https://github.com/moby/buildkit/issues/1850
+        name: Move cache
+        run: |
+          rm -rf /tmp/.binfmt-cache/${{ matrix.target }}
+          mv /tmp/.binfmt-cache/${{ matrix.target }}-new /tmp/.binfmt-cache/${{ matrix.target }}
+
+  publish:
+    runs-on: ubuntu-latest
+    needs: [ build-base, build ]
```
Comment (tonistiigi) [inline .github/workflows/ci.yml:100]: validate ?

Comment (crazy-max) [reply .github/workflows/ci.yml:100]: There is inheritance in place through `build` so I don't think it's required.

**Pull request merged**